### ISSUE TITLE
`license_finding_curations` with `source_artifact_url` do not get applied for packages

### ISSUE THREAD

**Issue by pepper-jk** [oss-review-toolkit/ort#7213]
Referencing a package via `source_artifact_url` in `package_configurations` results in `licensing_finding_curations` not to being applied, especially if defined in the `.ort.yml`.
It works with `vcs` information and fails with `source_artifact_url`.

I created a reproducible example [here](https://github.com/pepper-jk/ort_pypi_resolution_test/tree/pkg_curations) [1], which includes a `.ort.yml` with `package_configurations` that don't work and **out-commented** ones with `vcs` that do.
I even added `license_finding_curations` with explicit artifact paths and "project" `curations` to be sure, I did not mess up.

(Mapping `TCL` to `Vim` does not make any sense. It is just meant to be easily visible in the `evaluation-results.yml` and the `web-app-report`.)

### Additional Context:
Last week I complained about custom licenses not working for (package) curations.
After a lot of trail and error, I finally figured out that using `vcs` instead of `source_artifact_url` to reference the package did the trick. [2]

Now I have a different project where I have a package that does not have any `vcs` but only a source artifact. So now I need the `source_artifact_url` to actually work for `license_finding_curations` in `package_configurations`.

Upon feedback from @fviernau in Slack [3], I also tried creating a catch all curation, which also did not get applied:
```
package_configurations:
- id: "PyPI::pillow:9.2.0"
  source_artifact_url: "https://files.pythonhosted.org/packages/8c/92/2975b464d9926dc667020ed1abfa6276e68c3571dcb77e43347e15ee9eed/Pillow-9.2.0.tar.gz"
  license_finding_curations:
  - path: "**/*"
    reason: "INCORRECT"
    comment: "test curation"
    concluded_license: "Vim"
```

We then decided to open this issue.

[1] https://github.com/pepper-jk/ort_pypi_resolution_test/tree/pkg_curations
[2] https://ort-talk.slack.com/archives/C9NNJ54B1/p1687263011355689?thread_ts=1686827765.465319&cid=C9NNJ54B1
[3] https://ort-talk.slack.com/archives/C9NNJ54B1/p1687866704032119?thread_ts=1687865984.351029&cid=C9NNJ54B1

**Comment by sschuberth** [comment]
@pepper-jk, would you be able to contribute a test to `PackageConfigurationTest` that demonstrates the issue?

**Comment by pepper-jk** [comment]
I will try. :sweat_smile:

**Comment by pepper-jk** [comment]
Okay, I believe I understand the issue a bit better now. Let me explain what I learned so far:

The `PackageConfiguration.matches` function [1] only matches packages with same `ID` **and** `Provenance`, so the `Provenance` type must be the same as well. `ID` only is not enough. A package can have both **VCS** and **Source Artifact** information, but only one of them is chosen as `Provenance`.
Therefore, when crating a `curation` or at least `license_finding_curation` in a `package_configuration`, the `Provenance` information given needs to be precise.

Sadly this makes my example not particularly useful for my actual problem. As in the example, I try to match a `source_artifact_url` to a package with `vcs` `Provenance`, which is clearly not intended.
Quick note on this though: Maybe it should be more obvious which information is chosen as `Provenance`, as I have seen python packages choosing one over the other seemingly at random through different runs of ORT.

However, my actual problem is a package with no `Provenance` due to it not having a VCS, but only a `source_artifcat_url`, which was attached via a manually created SPDX file as I understand it from our customer, and does not seem to carry over as the `Provenance`. Need to take a deeper look into it, but that appears to be the problem.

I have to see if I can create a good example for it. Or even better code a test like you suggested. Unfortunately, the tests I wrote for the linked example seem of little value now. But I learned a lot.

So I guess the next step for me is diving into how a `Provenance` is assigned to a package and its license finding.

[1] https://github.com/oss-review-toolkit/ort/blob/main/model/src/main/kotlin/config/PackageConfiguration.kt#L73-L81

**Comment by fviernau** [comment]
@pepper-jk 

There is a `helper-cli` command which create a package configuration file for your.
If you use that, it should be ensured (if there is no bug) that the `id` and `source-artifact-url` it puts into the header actually do match. Wanna give it a try? (I'd recommend doing so via `orthw` see https://github.com/oss-review-toolkit/orthw)

**Comment by pepper-jk** [comment]
> There is a `helper-cli` command which create a package configuration file for your. If you use that, it should be ensured (if there is no bug) that the `id` and `source-artifact-url` it puts into the header actually do match. Wanna give it a try? (I'd recommend doing so via `orthw` see [oss-review-toolkit/orthw](https://github.com/oss-review-toolkit/orthw))

Thanks for the suggestion.

I have now tried to use the cli command briefly. Unfortunately, it does not seem to create any output file for the scan results I have of the project in question. Neither for the manually annotated `SpdxDocumentFile::triangle:1.6` nor for the automatically acquired `PyPI::paho-mqtt:1.6.1`.

<details>
<summary>orth package-configuration create output</summary>

```
# orth package-configuration create --scan-results-storage-dir /project/ort/all --package-id "PyPI::paho-mqtt:1.6.1" --output-dir /project/orth/
11:53:03,381 |-INFO in ch.qos.logback.classic.LoggerContext[default] - This is logback-classic version 1.4.7
11:53:03,397 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback-test.xml]
11:53:03,401 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback.xml] at [jar:file:/opt/ort/lib/cli-DOCKER-SNAPSHOT-740ea124cca8b06c141c5ea9d6ffbce7bf15994e.jar!/logback.xml]
11:53:03,402 |-WARN in ch.qos.logback.classic.util.DefaultJoranConfigurator@4ff4357f - Resource [logback.xml] occurs multiple times on the classpath.
11:53:03,402 |-WARN in ch.qos.logback.classic.util.DefaultJoranConfigurator@4ff4357f - Resource [logback.xml] occurs at [jar:file:/opt/ort/lib/helper-cli-DOCKER-SNAPSHOT-740ea124cca8b06c141c5ea9d6ffbce7bf15994e.jar!/logback.xml]
11:53:03,402 |-WARN in ch.qos.logback.classic.util.DefaultJoranConfigurator@4ff4357f - Resource [logback.xml] occurs at [jar:file:/opt/ort/lib/cli-DOCKER-SNAPSHOT-740ea124cca8b06c141c5ea9d6ffbce7bf15994e.jar!/logback.xml]
11:53:03,406 |-INFO in ch.qos.logback.core.joran.spi.ConfigurationWatchList@49cb9cb5 - URL [jar:file:/opt/ort/lib/cli-DOCKER-SNAPSHOT-740ea124cca8b06c141c5ea9d6ffbce7bf15994e.jar!/logback.xml] is not of type file
11:53:03,482 |-INFO in ch.qos.logback.core.model.processor.AppenderModelHandler - Processing appender named [STDOUT]
11:53:03,482 |-INFO in ch.qos.logback.core.model.processor.AppenderModelHandler - About to instantiate appender of type [ch.qos.logback.core.ConsoleAppender]
11:53:03,487 |-INFO in ch.qos.logback.core.model.processor.ImplicitModelHandler - Assuming default type [ch.qos.logback.classic.encoder.PatternLayoutEncoder] for [encoder] property
11:53:03,506 |-INFO in ch.qos.logback.classic.model.processor.RootLoggerModelHandler - Setting level of ROOT logger to WARN
11:53:03,507 |-INFO in ch.qos.logback.core.model.processor.AppenderRefModelHandler - Attaching appender named [STDOUT] to Logger[ROOT]
11:53:03,507 |-INFO in ch.qos.logback.classic.model.processor.LoggerModelHandler - Setting level of logger [org.apache.http.headers] to ERROR
11:53:03,507 |-INFO in ch.qos.logback.classic.model.processor.LoggerModelHandler - Setting level of logger [org.apache.http.wire] to ERROR
11:53:03,507 |-INFO in ch.qos.logback.classic.model.processor.LoggerModelHandler - Setting level of logger [org.eclipse.jgit.internal.storage.file.FileSnapshot] to ERROR
11:53:03,507 |-INFO in ch.qos.logback.classic.model.processor.LoggerModelHandler - Setting level of logger [org.ossreviewtoolkit.analyzer.managers.Yarn2] to INFO
11:53:03,507 |-INFO in ch.qos.logback.classic.model.processor.LoggerModelHandler - Setting level of logger [org.ossreviewtoolkit.clients.fossid.FossIdRestService] to INFO
11:53:03,507 |-INFO in ch.qos.logback.classic.model.processor.LoggerModelHandler - Setting level of logger [org.ossreviewtoolkit.reporter.reporters.fossid.FossIdReporter] to INFO
11:53:03,508 |-INFO in ch.qos.logback.classic.model.processor.LoggerModelHandler - Setting level of logger [org.ossreviewtoolkit.scanner.scanners.fossid.FossId] to INFO
11:53:03,508 |-INFO in ch.qos.logback.classic.model.processor.LoggerModelHandler - Setting level of logger [org.ossreviewtoolkit.scanner.scanners.fossid.FossIdConfig] to INFO
11:53:03,508 |-INFO in ch.qos.logback.classic.model.processor.LoggerModelHandler - Setting level of logger [org.ossreviewtoolkit.scanner.scanners.fossid.FossIdUrlProvider] to INFO
11:53:03,508 |-INFO in ch.qos.logback.core.model.processor.DefaultProcessor@55322aab - End of configuration.
11:53:03,508 |-INFO in ch.qos.logback.classic.joran.JoranConfigurator@2b4c1d96 - Registering current configuration as safe fallback point

```

</details>

I'm not sure if this is a problem with the project's scan results, the command or my usage of it. But either way I'm not sure I'll be able to investigate this due to time and access to the original project. I also don't have a setup of `orthw` ready to go. So have to pass on that for now, since I'm not sure it would work any better given the questionable input data.

There is of course a possibility this problem is related to the curations not being applied to this scan result. As I assume the mechanism to get the package in this `helper-cli` command are the same used during the curation process. But I'd rather debug the actual problem then the `helper-cli`, regardless if its the data or a bug in ORT. It might help me though to cross verify the behavior. Please let me know what you think.

Otherwise I'd concentrate on creating an reproducible example with `mqtt` and keep investigating the `Provenance` behavior of our problematic project.

**Comment by fviernau** [comment]
>  I also don't have a setup of orthw ready to go. 

I believe it's rather quick to setup (< 15 min.) and giving ROI quickly, but that's your call.
I only wanted to avoid wasting time due to a wrongly crafted file header, e.g. non matching id, source_artifact_url or vcs info. If you prefer another approach, up to you.

Creating the package config file with `orthw` would be as simple as `orth pc-create <PKG_ID>`

**Comment by pepper-jk** [comment]
I understand you are trying to minimize the margin of error for my package configuration. But keep in mind that for me using `orthw` only introduces more unknowns and steps to the problem.

However, the main reason I did not want to perform a setup was avoiding the corporate IT problems, one faces when setting something up in a VPNed and Proxied IT environment. But I got the go ahead to use my personal machine for this, so I might give it a try on Monday when I continue working on this.

Have a nice weekend.

**Comment by pepper-jk** [comment]
I have an update on my efforts. Sorry for the delay, I wanted to be certain first.

First of `orthw` was sadly no help. Since the project in question did not come with a vcs repo, it failed to `init` and therefore could not produce a package configuration. I did not spent any more time debugging it, since it was far from my original problem.

However, I solved my problem through tedious trial and error anyway. The problem was a mixture of misconfiguration and proxy issues. Unfortunately, whenever I had the correct config the proxies went haywire on me. Which was the main cause for me opening this issue. As I did review the evaluator results instead of the web app report during these tests, I missed the failed download of provenances. This of course meant the curations were not getting applied.

Afterwards, I switched from the VM with proxy problems to the GitLab pipeline. However, I had forgotten to push the `config.yml` with `enableRepositoryPackageConfigurations` [1]. So I again experienced the same problem for different reasons. After I fixed this I saw the first curations being applied. I'm happy to confirm that, package curations are now working, even with custom licenses and packages with `source_artifact_url`s only.

Some curations were trickier then others. Mainly the `starting_line` and `line_count` attribute were misaligned. To be exact they need to match the lines in the license finding 1:1, meaning they can not even be defined for a larger range, even if it includes the range of the finding. For more information on the topic see my Slack thread [2].

[1] https://github.com/oss-review-toolkit/ort/blob/main/docs/config-file-ort-yml.md#curating-project-license-findings
[2] https://ort-talk.slack.com/archives/C9NNJ54B1/p1689081352410399?thread_ts=1689081293.829809&cid=C9NNJ54B1
